active-rfcs/0000-finalize-head-body.md

```diff
@@ -0,0 +1,93 @@
+- **Out of scope / future:** The ability to author without `<body>`, `<head>`, or `<html>`, and relying on the compiler to handle them like the browser would.  This makes some sense because this is valid in HTML and we try to match HTML semantics whenever possible.  This is currently treated as out-of-scope, but there is nothing here that blocks our ability to do this in the future.
+  - For example, our compiler could automatically detect some elements in the component template (for example, a `<title>`) and auto-hoist it out of the component body and into the `<head>` of the final compiler result to support this.
+- **Out of scope / future:** Intelligent merging of repeated head elements. This is common when a top-level layout `<head>` might define a title, and then a more specific component might overwrite with a more specific `<title>` of its own. This is considered out of scope but not blocked to be implemented in the future.
+
+# Detailed design
+
+- `<head>`
+  - A `<head>` element can appear in any component, including pages, layouts, and simple UI components.
+  - The contents of a `<head>` element can be dynamic.
+  - Astro will collapse (flatten) all `<head>` elements together from the top-level page to the lowest component into the final page `<head>`.
+  - `<head>` elements are collected and and printed in the order that the components appear on the page.
+  - If no `<head>` elements are used by the author, no head will be output to the page.
+  - This solves for the page layout problem AND gives us new head management support, similar to [`<svelte:head>`](https://svelte.dev/docs#svelte_head)
+- `<body>`
+  - completely ignored in a component template
+  - a `<body>` tag is always included in the final output
+  - everything in the component template is assumed to be the body, other than the other elements listed here
+- `<html>`
+  - completely ignored in a component template
+  - an `<html>` tag is always included in the final output
+- `<!DOCTYPE html>`
+  - completely ignored in a component template
+  - a `<!DOCTYPE html>` tag is always included in the final output
+
+## Implementation
+
+Similar to how the Svelte compiler works, `head` elements would be collected in a `result.head` property by the compiler, seperate from the body of the template. 
+
+The runtime would collect these across all rendered components on the page, and at the end of the page render would have the final collapsed `head` to print to the page. 
+
+# Drawbacks
+
+- `<head>` ordering - I'm not sure if this is easy to implement. Could use a prototype or expert set of eyes :)
+- Does not support Streaming HTML for SSR. See "No `<head>` Collapsing" for an alternative approach that addresses this drawback. 
+
+# Alternatives
+
+## No `<head>` Collapsing
```

Comment:
@matthewp outlined a version of this proposal here that would remove the ability of any component to add elements to the head, in favor of supporting streaming HTML. I actually like the simplicity of this, and if this is easier to implement I might be more in favor of this for v1.0 (with the understanding that we could still add dynamic `<head>` injection later, post-v1.0 if there was enough interest)

Reply:
I prefer this to the current proposal, because it would be more _true_ to HTML, preventing automatic `<html>`, `<head>` or `<body>` generation from doing the wrong thing.